### PR TITLE
Lisa will no longer mate with corgi puppies. (Fixes istype checks in make_babies)

### PR DESCRIPTION
**Pull request by harmonyn**
morality has been restored to the world

The reason is actually relatively boring: whenever childtype was turned into lists, that fucked up the istype checks. Lisa and all simple animals were actually being prevented from mating with BYOND lists, not with their children.

The other thing affected by this is the max children in view of mob check, which also used istype incorrectly, which incorrectly allowed for the return of the petsplosion bug. Please do not bitch (no pun intended) about this. 

Fixes https://github.com/tgstation/tgstation/issues/41044.
Fixes https://github.com/tgstation/tgstation/issues/27550 for real.

### CODE REVIEW

**Comment by Barhandar** [comment]
Catsplosion is a feature, do not remove or readd if fixing unrelated issues.

**Comment by bgobandit** [comment]
@Barhandar by "catsplosion" I don't mean the wizard event

**Comment by Barhandar** [comment]
I explicitly mean Runtime being mated and resulting in CMO's office being filled with kittens - whether or not it's caused by her mating with said kittens, not wizard events.

**Comment by bgobandit** [comment]
This doesn't touch Runtime's persistence code. The only thing that is touched is a bug where the checks for istype(M, childtype) fail incorrectly because childtype is a list, not a path.

tl;dr: Please ACTUALLY READ THE PR.

**Comment by BebeYoshi** [comment]
I surely love when there are 3000 cats in the CMO office spamming emotes and even deleting them freezes the server xD

**Comment by obscolene** [comment]
Imagine unironically arguing about whether simplemobs should be able to commit pedophillic incest or not

**Comment by bgobandit** [comment]
I don't even give a fuck about that, this PR is literally just taking two lines of code that were objectively broken and making them not broken

**Review by SpaceManiac — CHANGES_REQUESTED** [review]
This won't be sufficient. For cats, `childtype = list(/mob/living/simple_animal/pet/cat/kitten)`, which means the typecache lookup will return `null`. The `childtype` value should probably be changed.

**Comment by bgobandit** [comment]
not sure why that would return null

**Comment by bgobandit** [comment]
not sure why that would return null

**Comment by Barhandar** [comment]
Bebe, that's why you bolt the office and drain the air instead.

File: code/modules/mob/living/simple_animal/simple_animal.dm [hunk 1]
```diff
@@ -377,12 +377,12 @@
...
 		if(M.stat != CONSCIOUS) //Check if it's conscious FIRST.
 			continue
-		else if(istype(M, childtype)) //Check for children SECOND.
+		else if(is_type_in_typecache(M, childtype)) //Check for children SECOND.
 			children++
 		else if(istype(M, animal_species))
 			if(M.ckey)
 				continue
-			else if(!istype(M, childtype) && M.gender == MALE) //Better safe than sorry ;_;
+			else if(!is_type_in_typecache(M, childtype) && M.gender == MALE) //Better safe than sorry ;_;
```
[the remark this answers is not in the document]
Comment (bgobandit) [reply code/modules/mob/living/simple_animal/simple_animal.dm]: there's a "better safe than sorry" comment so I assume there's a reason why there are two checks

[the remark this answers is not in the document]
Comment (ExcessiveUseOfCobblestone) [reply code/modules/mob/living/simple_animal/simple_animal.dm]: they used a `;_;` face so anything they say should be taken with a grain of cat litter.

**Comment by ExcessiveUseOfCobblestone** [comment]
"spam is ok if the code does it" will never be acceptable.

Also the lists aren't typecaches so does this work?

If not, 2 possible solutions are

`if(M.type in childtype)` (will be strict type and not consider subtypes)

OR

```DM
for(var/i in childtype)
    if(!istype(M,i))
        continue
    //rest of block here
```

**Comment by bgobandit** [comment]
this is tested and stops the mass reproduction, but I could change it to the second option (there are subtypes of kittens and puppies so the first won't work)

File: code/modules/mob/living/simple_animal/simple_animal.dm (hunk 1 above)
Comment (ShizCalev) [inline code/modules/mob/living/simple_animal/simple_animal.dm]: can lose this.

Comment (bgobandit) [reply code/modules/mob/living/simple_animal/simple_animal.dm]: see above "there's a "better safe than sorry" comment so I assume there's a reason why there are two checks"

**Comment by bgobandit** [comment]
@SpaceManiac I still don't know why cats would return null

**Comment by SpaceManiac** [comment]
```dm
#define is_type_in_typecache(A, L) (A && length(L) && L[(ispath(A) ? A : A:type)])
A = an instance of /mob/living/simple_animal/pet/cat/kitten
L = list(/mob/living/simple_animal/pet/cat/kitten)
(A && length(L) && L[(ispath(A) ? A : A:type)])
// A is not falsy:
(length(L) && L[(ispath(A) ? A : A:type)])
// L is not empty:
L[(ispath(A) ? A : A:type)]
// A is not a path:
L[A:type]
// A's type is /mob/living/simple_animal/pet/cat/kitten
L[/mob/living/simple_animal/pet/cat/kitten]
// there is no value in L associated with that key
null
```

**Comment by Cyberboss** [comment]
Your PR is getting old. Either Travis CI is failing, it is still labelled work in progress (change this by adding `[READY]` to the start of your title), or there are unaddressed change requests. This may be closed soon to free up space on our tracker if it remains idle. In the event that it is, please ask in #coderbus on Rizon for it to be reopened once you have resumed work on it.

**Comment by bgobandit** [comment]
will try to look at today

**Comment by bgobandit** [comment]
@SpaceManiac this still makes zero sense

**Comment by SpaceManiac** [comment]
try this in a test project:
```dm
/client/New()
    ..()
    var/list/L = list(/obj)
    src << isnull(L[/obj])  // will print 1
```

**Comment by optimumtact** [comment]
you guys should kiss and make up off tracker
